`@ConfigMapping` handling of Maps is not compatible with old config classes

### Describe the bug

For https://github.com/quarkusio/quarkus/issues/17664 I need to use `@ConfigMapping` as I don't want users to create deprecated config classes. Migration of the:

https://github.com/quarkusio/quarkus/blob/56ffa1223ca982c97933acdb08e6d6df4e5ca03d/extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerTenantConfig.java#L152

results in config validation error. Migrated class can be seen in here https://github.com/quarkusio/quarkus/commit/310a568eba3dfc856fd599b0fad147becdb5382e

### Expected behavior

There should be a way to do what we did in past without writing custom converted, that is:

```
@ConfigGroup
public static class ClaimInformationPointConfig {

    /**
     *
     */
    @ConfigItem(name = ConfigItem.PARENT)
    public Map<String, Map<String, Map<String, String>>> complexConfig;

    /**
     *
     */
    @ConfigItem(name = ConfigItem.PARENT)
    public Map<String, Map<String, String>> simpleConfig;
}
```


### Actual behavior

An exception is thrown and application starts to run:

```
io.smallrye.config.ConfigValidationException: Configuration validation failed:
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "static-claim")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "{request.parameter['grant']}")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "GET")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "{request.body['/from-body']}")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "{request.relativePath}")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "static-claim")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "static-claim")
	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:643)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:727)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: io.smallrye.config.ConfigValidationException: Configuration validation failed:
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "static-claim")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "{request.parameter['grant']}")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "GET")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "{request.body['/from-body']}")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "{request.relativePath}")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "static-claim")
	java.util.NoSuchElementException: SRCFG00042: Value does not match the expected map format "<key1>=<value1>;<key2>=<value2>..." (value was "static-claim")
	at io.smallrye.config.ConfigMappingProvider.mapConfiguration(ConfigMappingProvider.java:75)
	at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:84)
	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:714)
	at io.quarkus.deployment.configuration.BuildTimeConfigurationReader.initConfiguration(BuildTimeConfigurationReader.java:409)
	at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:137)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:107)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:330)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:251)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:60)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:224)
	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:610)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:660)
	... 1 more
```

### How to Reproduce?

Steps to reproduce the behavior:

1. `git@github.com:michalvavrik/quarkus.git`
2. `cd quarkus`
3. `git checkout feature/kc-authz-dynamic-policy-enforcer-resolver`
4. `mvn -Dquickly`
5. `mvn clean verify -f integration-tests/keycloak-authorization/ -Dtest-containers -Dstart-containers`

### Output of `uname -a` or `ver`

Fedora 38

### Output of `java -version`

21

### Quarkus version or git rev

999-SNAPSHOT

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.3

### Additional information

_No response_

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -51,7 +51,7 @@
         <microprofile-lra.version>2.0</microprofile-lra.version>
         <microprofile-openapi.version>3.1.1</microprofile-openapi.version>
         <smallrye-common.version>2.3.0</smallrye-common.version>
-        <smallrye-config.version>3.6.1</smallrye-config.version>
+        <smallrye-config.version>3.7.0</smallrye-config.version>
         <smallrye-health.version>4.1.0</smallrye-health.version>
         <smallrye-metrics.version>4.0.0</smallrye-metrics.version>
         <smallrye-open-api.version>3.10.0</smallrye-open-api.version>

--- a/devtools/gradle/gradle/libs.versions.toml
+++ b/devtools/gradle/gradle/libs.versions.toml
@@ -3,7 +3,7 @@ plugin-publish = "1.2.1"
 
 # updating Kotlin here makes QuarkusPluginTest > shouldNotFailOnProjectDependenciesWithoutMain(Path) fail
 kotlin = "1.9.23"
-smallrye-config = "3.6.1"
+smallrye-config = "3.7.0"
 
 junit5 = "5.10.2"
 assertj = "3.25.3"

